src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                case DecisionTree.DecisionKind.Guarded:
+                    return AddByValue((DecisionTree.Guarded)decision, value, makeDecision, hasErrors);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(decision.Kind);
+            }
+        }
+
+        private DecisionTree AddByValue(DecisionTree.Guarded guarded, BoundExpression value, DecisionMaker makeDecision, bool hasErrors)
+        {
+            if (guarded.Default != null)
+            {
+                if (guarded.Default.MatchIsComplete) return null;
+            }
+            else
+            {
+                guarded.Default = new DecisionTree.ByValue(guarded.Expression, guarded.Type);
+                guarded.Default.Parent = guarded;
+            }
+
+            return AddByValue(guarded.Default, value, makeDecision, hasErrors);
+        }
+
+        private DecisionTree AddByValue(DecisionTree.ByValue byValue, BoundExpression value, DecisionMaker makeDecision, bool hasErrors)
+        {
+            Debug.Assert(value.Type == byValue.Type);
+            if (byValue.Default != null)
+            {
+                return AddByValue(byValue.Default, value, makeDecision, hasErrors);
+            }
+
+            if (byValue.Expression.ConstantValue != null)
+            {
+                // We are switching on a constant expression. See if the constant value matches
+                bool matches = byValue.Expression.ConstantValue == value.ConstantValue;
+                if (!matches) return null;
+                byValue.Default = makeDecision(byValue.Expression, byValue.Type);
+                byValue.Default.Parent = byValue;
+                if (byValue.Default.MatchIsComplete)
+                {
+                    byValue.MatchIsComplete = true;
+                }
+                return byValue.Default;
+            }
+
+            // For error recovery, to avoid "unreachable code" diagnostics when there is a bad case
+            // label, we use the case label itself as the value key.
+            object valueKey = hasErrors ? (object)value : value.ConstantValue.Value;
+            DecisionTree valueDecision;
+            if (byValue.ValueAndDecision.TryGetValue(valueKey, out valueDecision))
+            {
+                return Add(valueDecision, makeDecision);
+            }
+            valueDecision = makeDecision(byValue.Expression, byValue.Type);
+            valueDecision.Parent = byValue;
+            byValue.ValueAndDecision.Add(valueKey, valueDecision);
+            return valueDecision;
+        }
+
+        private DecisionTree AddByValue(DecisionTree.ByType byType, BoundExpression value, DecisionMaker makeDecision, bool hasErrors)
+        {
+            if (byType.Default != null)
+            {
+                try
+                {
+                    return AddByValue(byType.Default, value, makeDecision, hasErrors);
+                }
+                finally
```

Comment:
I don't know what you mean by "needed". This is a correct implementation of the intended logic. There are other ways to express the logic as well. I do not expect exceptions.

---
In reply to: [64977864](https://github.com/dotnet/roslyn/pull/11485#discussion_r64977864) [](ancestors = 64977864)